react/features/whiteboard/middleware.ts

```diff
@@ -0,0 +1,201 @@
+/* eslint-disable import/order */
+import { generateCollaborationLinkData } from '@jitsi/excalidraw';
+import { IStore } from '../app/types';
+
+// @ts-ignore
+import { setTileView } from '../video-layout/actions.any';
+import { PARTICIPANT_UPDATED, PIN_PARTICIPANT } from '../base/participants/actionTypes';
+import { getLocalParticipant, isEveryoneModerator, isParticipantModerator } from '../base/participants/functions';
+import { participantJoined, participantLeft, pinParticipant } from '../base/participants/actions';
+
+// @ts-ignore
+import { getCurrentConference } from '../base/conference/functions';
+import { CONFERENCE_JOIN_IN_PROGRESS } from '../base/conference/actionTypes';
+
+// @ts-ignore
+import { MiddlewareRegistry, StateListenerRegistry } from '../base/redux';
+import { HIDE_WHITEBOARD, OPEN_WHITEBOARD, SHOW_WHITEBOARD } from './actionTypes';
+import { isWhiteboardOpen, isWhiteboardPresent } from './functions';
+import { WHITEBOARD, WHITEBOARD_ID, WHITEBOARD_PARTICIPANT_NAME, WHITEBOARD_STORAGE_KEY } from './constants';
+import { closeWhiteboard, setupWhiteboard } from './actions';
+import { getCurrentRoomId } from '../breakout-rooms/functions';
+
+// @ts-ignore
+import { addStageParticipant, removeStageParticipant } from '../filmstrip/actions.web';
+
+// @ts-ignore
+import { isStageFilmstripAvailable } from '../filmstrip/functions';
+
+declare let APP: any;
+
+const focusWhiteboard = (store: IStore) => {
+    const { dispatch, getState } = store;
+    const state = getState();
+    const conference = getCurrentConference(state);
+    const stageFilmstrip = isStageFilmstripAvailable(state);
+    const isPresent = isWhiteboardPresent(state);
+
+    if (!isPresent) {
+        dispatch(participantJoined({
+            conference,
+            id: WHITEBOARD_ID,
+            isFakeParticipant: true,
+            isWhiteboard: true,
+            name: WHITEBOARD_PARTICIPANT_NAME
+        }));
+    }
+    if (stageFilmstrip) {
+        dispatch(addStageParticipant(WHITEBOARD_ID, true));
+    } else {
+        dispatch(pinParticipant(WHITEBOARD_ID));
+    }
+};
+
+const blurWhiteboard = (store: IStore) => {
+    const { dispatch, getState } = store;
+    const state = getState();
+    const stageFilmstrip = isStageFilmstripAvailable(state);
+
+    if (stageFilmstrip) {
+        dispatch(removeStageParticipant(WHITEBOARD_ID));
+    }
+};
+
+/**
+ * Handles the whiteboard state.
+ *
+ * @param {Store} store - The redux store.
+ * @param {Object} attributes - The command attributes.
+ * @returns {void}
+ */
+const handleWhiteboardStatus = (store: IStore, { roomId, roomKey }: { roomId: string; roomKey: string; }) => {
+    const { dispatch, getState } = store;
+    const state = getState();
+    const username = getLocalParticipant(state)?.name;
+    const isOpen = isWhiteboardOpen(state);
+
+    if (isOpen) {
+        return;
+    }
+
+    localStorage.setItem(WHITEBOARD_STORAGE_KEY, JSON.stringify({ username }));
```

Comment:
Now that I think of it, why are we doing this, if we are passing our display name to the component via the API?

Reply:
I don't understand the question, whose API? Excalidraw needs to take the collaborator username from the local storage.

Reply:
But we also use the API to update it when we change the local display name, don't we?

Reply:
Oh I was under the impression that the API calls the same action (PARTICIPANT_UPDATED). This part here is intended for when a participant is joining the whiteboard for the first time, so I think it needs to stay. To cover the API I should update the local storage from here too https://github.com/jitsi/jitsi-meet/blob/master/modules/API/API.js#L193 right?

Reply:
NO, I mean the collab API here. We can have the useEffect run on the first render too, so the name is set, can't we?

Reply:
I tried it out and we still can't do without adding the initial username to local storage, because on the first render the collabAPI is not yet available and the localUsername is already set, so until the next username update, excalidraw won't be notified of what to display on the cursor.

Reply:
I don't think this is not solvable. We do have the name in Redux before the render.

Reply:
Solved, thank you!